crates/block-producer/src/mempool/batch_graph.rs

```diff
@@ -0,0 +1,209 @@
+            // call as well.
+            for child in node.children {
+                // Its possible for the child to part of this same set of batches and therefore
+                // already removed.
+                if let Some(child) = self.nodes.get_mut(&child) {
+                    child.parents.remove(&batch);
+                }
+            }
+
+            transactions.extend_from_slice(&node.transactions);
+        }
+
+        transactions
+    }
+
+    /// Mark a batch as proven if it exists.
+    pub fn mark_proven(&mut self, id: BatchJobId, batch: TransactionBatch) {
+        // Its possible for inflight batches to have been removed as part of another batches
+        // failure.
+        if let Some(node) = self.nodes.get_mut(&id) {
+            node.status = Status::Proven(batch);
+            self.try_make_root(id);
+        }
+    }
+
+    /// Returns at most `count` __indepedent__ batches which are ready for inclusion in a block.
+    pub fn select_block(&mut self, count: usize) -> BTreeMap<BatchJobId, TransactionBatch> {
+        let mut batches = BTreeMap::new();
+
+        // Track children so we can evaluate them for root afterwards.
+        let mut children = BTreeSet::new();
+
+        for batch_id in &self.roots {
+            let mut node = self.nodes.get_mut(batch_id).expect("Root node must be in graph");
+
+            // Filter out batches which have dependencies in our selection so far.
+            if node.parents.iter().any(|parent| batches.contains_key(parent)) {
+                continue;
+            }
+
+            let Status::Proven(batch) = node.status.clone() else {
+                unreachable!("Root batch must be in proven state.");
+            };
+
+            batches.insert(*batch_id, batch);
+            node.status = Status::InBlock;
+
+            if batches.len() == count {
+                break;
+            }
+        }
+
+        // Performing this outside the main loop has two benefits:
+        //   1. We avoid checking the same child multiple times.
+        //   2. We avoid evaluating children for selection (they're dependents).
+        for child in children {
+            self.try_make_root(child);
+        }
+
+        batches
+    }
+
+    fn try_make_root(&mut self, id: BatchJobId) {
+        let node = self.nodes.get_mut(&id).expect("Node must be in graph");
+
+        for parent in node.parents.clone() {
+            let parent = self.nodes.get(&parent).expect("Parent must be in pool");
+
+            if parent.status != Status::InBlock {
+                return;
+            }
+        }
+        self.roots.insert(id);
+    }
+}
+
+#[derive(Clone, Debug)]
+struct Node {
+    status: Status,
+    transactions: Vec<TransactionId>,
+    parents: BTreeSet<BatchJobId>,
+    children: BTreeSet<BatchJobId>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+enum Status {
+    InFlight,
+    Proven(TransactionBatch),
+    InBlock,
+}
```

Comment:
Would be good to add comments explaining what these enum variants mean. Also, to keep them somewhat parallel with transaction status, maybe the names could be `Queued`, and `Proven`, and `Processed`?